### ISSUE TITLE
NaNs in torch.nn.MultiheadAttention output with certain params

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#81077]
### Summary 
While experimenting with different sets of hyperparamters (batch_size, max_sequence_len, embed_size) for `torch.nn.MultiheadAttention` I found that for certain tuples the forward of this module would produce NaNs. 

### A minimal example demonstrating this:

``` Python
import torch
import random
import numpy as np

import warnings

warnings.filterwarnings("ignore")

@torch.inference_mode()
def main():
    random.seed(42)
    torch.manual_seed(42)
    np.random.seed(42)

    a, b = 1, 0.8
    batch_size = 128
    sequence_length = 128
    block_size = 32
    embed_dim = 1024
    num_heads = 16
    device = "cuda"
    dtype = torch.float16

    from scipy.stats import beta
    # lengths is ultimately just a list of numbers in the range [0,sequence_length] where each number is a multiple of block_size
    lengths = beta.rvs(a, b, size=batch_size) * (sequence_length + block_size) // block_size
    lengths_in_blocks = list(map(int, list(lengths)))
    lengths = [l * block_size for l in lengths_in_blocks]

    # Create q,k,v
    q = [torch.randn(l, embed_dim, device=device, dtype=dtype)for l in lengths]
    q = torch.nested_tensor(q, device=device, dtype=dtype)
    k, v = q, q

    qkv = torch.nn.Linear(embed_dim, 3 * embed_dim, device=device, dtype=dtype)
    proj = torch.nn.Linear(embed_dim, embed_dim, device=device, dtype=dtype)

    native_mha = torch.nn.MultiheadAttention(embed_dim, num_heads, batch_first=True, device=device, dtype=dtype).eval()
    native_mha.in_proj_weight = qkv.weight
    native_mha.in_proj_bias = qkv.bias
    native_mha.out_proj.weight = proj.weight
    native_mha.out_proj.bias = proj.bias

    y_native_mha, *rest = native_mha(q,k,v, need_weights=False)
    assert not torch.isnan(y_native_mha.to_padded_tensor(0)).any(), "I wouldn't expect NaNs to be present"

if __name__ == "__main__":
    main()
```

I thought this might be related to using float16 but I also tried float32 and float64 which all produced NaNs. There are other *magic* sets of batch_size, max_sequence_length, num_heads, embed_dim that produce NaNs. If it is worth enumerating a few more I can gather them.

### Versions

```
PyTorch version: 1.13.0a0+git8ba15ee
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.22.5
Libc version: glibc-2.27

Python version: 3.9.12 (main, Jun  1 2022, 11:38:51)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-1069-aws-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 11.6.112
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 510.47.03
cuDNN version: Probably one of the following:
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.1.1
/usr/local/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

Versions of relevant libraries:
[pip3] mypy==0.950
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] torch==1.13.0a0+git8ba15ee
[conda] blas                      1.0                         mkl
[conda] magma-cuda116             2.6.1                         0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-include               2022.0.1           h06a4308_117
[conda] mkl-service               2.4.0            py39h7f8727e_0
[conda] mkl_fft                   1.3.1            py39hd3c417c_0
[conda] mkl_random                1.2.2            py39h51133e4_0
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] torch                     1.13.0a0+git8ba15ee          pypi_0    pypi
```

cc @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by drisspg** [comment]
First pattern that jumps out to me is when batch_size and max_sequence length are equal this problem occurs.
I also tried batch_size 512 and max_sequence_length of 512 and this also produced NaNs.... now to figure out why
```
{
'batch:64_seq_len:64_n_heads:16_embed_dim:512': 121, 
'batch:64_seq_len:64_n_heads:16_embed_dim:1024': 121, 
'batch:128_seq_len:128_n_heads:16_embed_dim:512': 121, 
'batch:128_seq_len:128_n_heads:16_embed_dim:1024': 121, 
'batch:256_seq_len:256_n_heads:16_embed_dim:512': 121,
'batch:256_seq_len:256_n_heads:16_embed_dim:1024': 121
}
```

**Comment by erichan1** [comment]
Good callout! Did some digging. When I print the lengths for the example code I see that there are some zero length sequences
```
[96, 0, 0, 128, 96, 64, 96, 128, 64, 32, 128, 128, 128, 32, 0, 32, 64, 128, 64, 0, 96, 32, 0, 128, 0, 128, 128, 128, 64, 0, 96, 32, 0, 64, 128, 96, 128, 128, 32, 0, 96, 32, 96, 128, 32, 32, 32, 32, 0, 96, 0, 128, 96, 128, 96, 128, 32, 96, 0, 64, 128, 128, 0, 0, 128, 128, 32, 96, 64, 32, 128, 128, 0, 64, 32, 128, 32, 96, 64, 128, 128, 64, 32, 96, 32, 128, 128, 0, 96, 128, 128, 0, 96, 64, 96, 96, 0, 96, 128, 64, 32, 64, 128, 64, 128, 0, 0, 64, 64, 128, 128, 0, 32, 128, 0, 96, 0, 128, 128, 96, 64, 0, 0, 128, 0, 128, 0, 0]```

I was able to fix it by forcing all sequence lengths to be at least size 1. 

```
@torch.inference_mode()
def test_transformerencoder_nans(force_seqlen_nonzero=False):
    random.seed(42)
    torch.manual_seed(42)
    np.random.seed(42)

    a, b = 1, 0.8
    batch_size = 128
    sequence_length = 128
    block_size = 32
    embed_dim = 1024
    num_heads = 16
    device = "cuda"
    dtype = torch.float16

    from scipy.stats import beta
    # lengths is ultimately just a list of numbers in the range [0,sequence_length] where each number is a multiple of block_size
    lengths = beta.rvs(a, b, size=batch_size) * (sequence_length + block_size) // block_size
    lengths_in_blocks = list(map(int, list(lengths)))
    lengths = [l * block_size for l in lengths_in_blocks]
    if(force_seqlen_nonzero):
        for i in range(len(lengths)):
            lengths[i] = max(lengths[i], 1)
    print(lengths)

    # Create q,k,v
    q = [torch.randn(l, embed_dim, device=device, dtype=dtype)for l in lengths]
    q = torch.nested_tensor(q, device=device, dtype=dtype)
    k, v = q, q

    native_mha = torch.nn.MultiheadAttention(embed_dim, num_heads, batch_first=True, device=device, dtype=dtype).eval()

    y_native_mha, *rest = native_mha(q,k,v, need_weights=False)
    assert not torch.isnan(y_native_mha.to_padded_tensor(0)).any(), "I wouldn't expect NaNs to be present"
```

**Comment by drisspg** [comment]
Ohh nice catch I had been wondering what happens when passing in a 0 len sequence in batch. I would assume(hope) this never really happens in real data.  I am running through the full sweep now with force_seqlen_nonzero functionality to make sure  they all work.  One slightly strange thing is that the NaN issue w/ 0 only seems to occur when the the max_seq_len == batch_size. If you make the ranges 1 bigger in either direction even with 0 then no NaNs.

**Comment by drisspg** [comment]
Hmm so while this does remove the NaNs the output form native_mha and flash_attention are deviating for the cases where Batch_Size == max_seq_len.
```
torch.testing.assert_close(y_native_mha, y_internal_flash, atol=1e-3, rtol=1e-3)
Tensor-likes are not close!

Mismatched elements: 457182 / 2097152 (21.8%)
Greatest absolute difference: 0.19415283203125 at index (32, 16, 296) (up to 0.001 allowed)
Greatest relative difference: 302812.4285714286 at index (41, 7, 411) (up to 0.001 allowed)
```

Previously this was just showing NaN comparisons for the reason for the differences now it is finite values but still different

**Comment by cpuhrsch** [comment]
We've been having issues with nn.MHA producing NaNs when passed fully masked inputs for a while now. [MaskedTensors](https://pytorch.org/maskedtensor/main/index.html) should be able to address this, by allowing us to return a fully masked out row when given a fully masked out input. Otherwise it's definitely possible to have NestedTensors with 0-sized sizes across irregular dimensions.

**Comment by drisspg** [comment]
Closing because there doesn't seem to be much activity and no longer observe this behavior on local tests.
